### ISSUE TITLE
Is there any way to remove the 1px border?

### ISSUE THREAD

**Issue by Jasper0819X** [melak47/BorderlessWindow#13]
My recent project requires a semi-transparent borderless window. I'm using your method and direct2d for rendering on window 10:
![image](https://cloud.githubusercontent.com/assets/15362491/21890160/5dfba4ca-d907-11e6-8670-d3f39d6916ea.png)
(set alpha = 0.5F)
![image](https://cloud.githubusercontent.com/assets/15362491/21890224/a547e8c0-d907-11e6-80c0-f2369e1148bd.png)

(set alpha = 0.0F)

As you can see there is still a 1px border. Is there any way to remove it?

**Comment by melak47** [comment]
Unfortunately I have come to the conclusion that it's not possible to paint over the one pixel border on Windows 10, see issue #11 if you care for some more details.

However, *if you don't require the soft shadow*, you can just use the `basic_borderless` window style (`WS_POPUP` derived) instead of `aero_borderless`, which does not come with such a border. 
It should still work with the glass/"blur behind" effect:

![basic_borderless](https://cloud.githubusercontent.com/assets/3340026/21903971/2e8954ee-d902-11e6-9b42-51d1fe4e3ff9.png)

**Comment by melak47** [comment]
@Jasper0819X By the way, how are you enabling the blur-behind?
In your screenshots the blur only appears in the client area, 
but if I use the undocumented `SetWindowCompositionAttribute` function, the blur extends beyond the 1px border.

**Comment by ZimM-LostPolygon** [comment]
Yes! I've found a way to have both the shadow and get rid of 1px border on Window 10. Unfortunately, it relies on undocumented `SetWindowCompositionAttribute` function. Here's the reference:
https://gist.github.com/riverar/fd6525579d6bbafc6e48
The AccentState has to be set to `ACCENT_ENABLE_GRADIENT`.

**Comment by ZimM-LostPolygon** [comment]
@melak47 
However... I've just tested your sample in Windows 7 (just made the background black), and the 1 px border is there as well. And it is just transparent, which looks weird. Am I missing something? I believed the issue was only with Windows 10.
![default](https://user-images.githubusercontent.com/928550/27249642-b1886bf4-531a-11e7-9be0-73876202aa51.png)

**Comment by melak47** [comment]
@ZimM-LostPolygon the 1 px border you see is the result of calling `DwmExtendFrameIntoClientArea`. It's there on both Windows 7 and Windows 10, but for some reason Windows 10 paints over it using your background brush, while Windows 7 will not.
You can paint over it yourself (e.g. using D2D, D3D, OGL, ...) -- *but* you have to use an opaque color -- @Jasper0819X wants to use some level of transparency for the blur behind effect to show, and that's why the border is showing through. 
Here's an example of going from fully opaque to fully transparent (with blur behind enabled):
![capture](https://user-images.githubusercontent.com/3340026/27255502-674d65aa-539f-11e7-981c-2d9afeb0984d.gif)

So as long as you can put something opaque in your client area you should be ok.

**Comment by ZimM-LostPolygon** [comment]
@melak47 
Hm... I am painting on a window using D3D, but the transparent border is still there on Windows 7 as well. Do I have to draw with some special coordinates or something?

**Comment by melak47** [comment]
No special coordinates should be needed.
Do you clear the backbuffer with an opaque color (say RGBA (0,0,0,1))?

**Comment by melak47** [comment]
@ZimM-LostPolygon I just tried filling the window using D2D on Windows 7, and it works fine for me:
![d2d](https://user-images.githubusercontent.com/3340026/27256384-5c3b07da-53b3-11e7-9f74-a79129fb52fa.png)
The code is here: https://github.com/melak47/BorderlessWindow/tree/d2d-fill if you want to compare.
No need to use `SetWindowCompositionAttribute`.

**Comment by ZimM-LostPolygon** [comment]
@melak47 
Ah, indeed, I was filling the window with RGBA(0,0,0,0). Making the color fully opaque fixed it on Windows 7.

BTW, my Windows 10 solution that uses the `SetWindowCompositionAttribute` function avoids this problem completely, since `DwmExtendFrameIntoClientArea` is not called at all. The only downside is that shadow is always the same shadow that is used for non-active windows. But for me this is better than the 1px border. Also, there shouldn't be any problems with controlling the opacity. Here's how it looks:
![default](https://user-images.githubusercontent.com/928550/27256531-463d6c7a-53b8-11e7-98e3-354321177046.png)

**Comment by ZimM-LostPolygon** [comment]
@melak47 
`SetWindowCompositionAttribute` trick is only for Windows 10. The `AccentPolicy` and related APIs are only available there. The 1px border on Windows 10 *is* the accent border.

**Comment by melak47** [comment]
@ZimM-LostPolygon I can't quite reproduce your results. If I use `ACCENT_ENABLE_GRADIENT` without `DwmExtendFrameIntoClientArea` as you suggest, I get no aero shadow at all. (Also, now I can't use `ACCENT_ENABLE_BLURBEHIND` anymore, so what opacity is there to control?)

Aren't you going to draw to the window, anyway? Why is clearing with an opaque color a problem? :)

**Comment by ZimM-LostPolygon** [comment]
Well, it is a problem, since in the end, I want to actually *use* the alpha channel for transparent effects :)

Here's the code snippet in C# that I used for the screenshot:
```
WinApi.DwmApi.Undocumented.AccentPolicy accent = new WinApi.DwmApi.Undocumented.AccentPolicy();
accent.AccentState = WinApi.DwmApi.Undocumented.AccentState.ACCENT_ENABLE_BLURBEHIND;
accent.AccentState = WinApi.DwmApi.Undocumented.AccentState.ACCENT_ENABLE_GRADIENT;
//accent.AccentState = WinApi.DwmApi.Undocumented.AccentState.ACCENT_ENABLE_TRANSPARENTGRADIENT;
accent.AccentFlags = (WinApi.DwmApi.Undocumented.AccentFlags) (0x20 | 0x40 | 0x80 | 0x100);
// Uncomment next line to disable shadows
//accent.AccentFlags = 0;
accent.GradientColor = 0;

int accentStructSize = Marshal.SizeOf(accent);
IntPtr accentPtr = Marshal.AllocHGlobal(accentStructSize);
Marshal.StructureToPtr(accent, accentPtr, false);

WinApi.DwmApi.Undocumented.WindowCompositionAttributeData data = new WinApi.DwmApi.Undocumented.WindowCompositionAttributeData();
data.Attribute = WinApi.DwmApi.Undocumented.WindowCompositionAttribute.WCA_ACCENT_POLICY;
data.SizeOfData = accentStructSize;
data.Data = accentPtr;

WinApi.DwmApi.Undocumented.SetWindowCompositionAttribute(hWnd, ref data);
Marshal.FreeHGlobal(accentPtr);
```
As for the opaque color... Well, initially, I wanted a window with both blur-behind and shadows. Using `SetWindowCompositionAttribute` with `ACCENT_ENABLE_BLURBEHIND` allows for that:
![default](https://user-images.githubusercontent.com/928550/27260019-450430c8-5421-11e7-8072-fbb800efef62.png)
The gray 1px border you can notice here is actually part of the shadow, it is half-transparent, so it looks fine on all background. And the opacity control using `SetLayeredWindowAttributes` works correctly, even without shadows:
![default](https://user-images.githubusercontent.com/928550/27260064-0c833bde-5423-11e7-8865-a2632a9957af.png)

If I don't want the blur-behind, but still want to have working transparency without the 1px accent color border, I can use `ACCENT_ENABLE_TRANSPARENTGRADIENT` with `AccentPolicy.GradientColor` set to `0x01000000` and get this:
![default](https://user-images.githubusercontent.com/928550/27260101-df3756d2-5423-11e7-9724-4d3fb2cc2f72.png)
So basically... Using `SetWindowCompositionAttribute` removes the need to paint over with an opaque color, opening some new possibilities on Windows 10.

**Comment by melak47** [comment]
Ah, those are the flags used to draw the borders on the start menu and the like.
```
enum AccentFlag {
	DrawLeftBorder   = 1 << 5,
	DrawTopBorder    = 1 << 6,
	DrawRightBorder  = 1 << 7,
	DrawBottomBorder = 1 << 8,
	DrawAllBorders   = (DrawLeftBorder | DrawTopBorder | DrawRightBorder | DrawBottomBorder),
};
```
I've seen someone write about these before, but I hadn't found a good use for them, thanks.

**Comment by Jasper0819X** [comment]
I rebuilt my project on Windows Build 1703 with codes below:
```
static const MARGINS shadow_state{ 0,0,0,0 };
::DwmExtendFrameIntoClientArea(handle_, &shadow_state);
```

the border just disappeared and works fine with direct2d:
![image](https://user-images.githubusercontent.com/15362491/30047306-084f0e0e-9243-11e7-8dd8-7ba4774e6a01.png)

**Comment by mezstd** [comment]
This way seems to work:

`   ::SetWindowLongPtr(hWnd, GWL_EXSTYLE, ::GetWindowLongPtr(hWnd, GWL_EXSTYLE) | WS_EX_LAYERED);
   ::SetLayeredWindowAttributes(hWnd, 0x80808080, 255, LWA_COLORKEY | LWA_ALPHA);`
